### PR TITLE
Introduce qimalloc as an allocator option

### LINKED ISSUE #46: Implement "primitive allocator" which never frees
This is (probably) useful for precompiles.

### PR DESCRIPTION
**Pull request by s1na**
Closes #46.

So far it's a really dumb allocator. It simply tries to allocate enough wasm pages everytime the `alloc` method is called (i.e. pages are not filled with subsequent alloc calls). It also doesn't dealloc. I tested it with `runevm+hera` and it passes some tests.

### CODE REVIEW

**Comment by s1na** [comment]
Some notes:

- The allocation algorithm is a rough translation of @poemm's C [malloc](https://github.com/poemm/C_ewasm_contracts/blob/a3276b1242c22f275862869572e77104f1895974/src/ewasm.h#L128)
- It's been only implemented for arch `wasm32`
- I tried it with a few simple test cases, but not fully sure about correctness

From a rust developer PoV this code is probably an abomination :smile:

**Comment by poemm** [comment]
The C version of dumb `malloc()` is untested for edge cases and assumes things about LLVM compiler output.

The good news: dumb `malloc()` compiles to wasm in 15 bytes (or 60 bytes if the preprocessor flag `GROWABLE_MEMORY` is set to `true`). This mostly solves the problem of memory allocation in Ewasm -- no host system library is needed for `malloc()` since its size is nearly negligible.

**Comment by axic** [comment]
Changed this to use the new crate: https://github.com/wasmx/qimalloc

**Comment by s1na** [comment]
Looks good to me, can't approve my own PR though :)

File: src/lib.rs [hunk 1]
```diff
@@ -18,10 +18,21 @@
...
 /// }
 /// ```
-extern crate wee_alloc;
 
-#[global_allocator]
-static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+#[macro_use]
+extern crate cfg_if;
+
+cfg_if! {
```
Comment (jakelang) [inline src/lib.rs]: I am pretty sure we don't need the `cfg_if` dependency here. We can do something like `#[cfg(feature = "wee_alloc")]  static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;`

Comment (axic) [reply src/lib.rs]: It also has/needs that `#[global_allocator]` annotation. Will that work?

Comment (jakelang) [reply src/lib.rs]: I think so, attributes apply to the item below. If the item is not compiled in, then the attribute isn't either.

**Comment by axic** [comment]
Release qimalloc so we can refer to a version and not just a random repo in this crate.
